### PR TITLE
*: provide virtual table facilities

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Inspired by the hackathon2018 project -- [TiNiuB](https://github.com/TiNiuB/tiquery)

This PR aimed to provide some basic virtual table facilities, so we incubate the idea from the hackathon further.

The ideal usage would like that:

```
import _ "github.com/TiNiuB/tiquery/driver"   // provide the TiNiuB virtual table implementation
```

Implement some virtual tables in a Go package, and import that package register the tables in TiDB automatically. And the package could be implement as a plugin if possible, that's really neat.

### What is changed and how it works?

* Modify the `ddl` package to provide the `BuildTableInfoFromSQL`
* Modify the `infoschema` package to provide the `RegisterVirtualTable` mechanism
* Refactor the `perfschema` package to serve as an example

### Check List 

Tests 

 - No code

Basically, no new things are introduced

Code changes

 - Has exported function/method change

@winkyao @lysu @jackysp

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8657)

### CODE REVIEW

File: perfschema/init.go [hunk 1]
```diff
@@ -14,344 +14,50 @@
...
-		meta.ID = autoid.GenLocalSchemaID()
-		for _, c := range meta.Columns {
-			c.ID = autoid.GenLocalSchemaID()
+// Init register the PERFORMANCE_SCHEMA virtual tables.
+// It should be init(), and the ideal usage should be:
+//
+// import _ "github.com/pingcap/tidb/perfschema"
+//
+// This function depends on plan/core.init(), which initialize the expression.EvalAstExpr function.
+// The initialize order is a problem if init() is used as the function name.
```
Comment (lysu) [inline perfschema/init.go]: should we add a `/pkgdep` package in central control  to solve init problem, there are not benefit to use lazy `init()` call in almost situation.

so how about will init pkg that need init ahead.

1. add `github.com/pingcap/tidb/pkgdep` , and add a file `initdep.go`

```
package pkgdep

import (
	
	_ "github.com/pingcap/tidb/planner/core"

	_ "github.com/pingcap/tidb/perfschema"

)
```
2. `in main.go`  just import pkgdep first

```
import (
	"flag"
	"fmt"
	"os"
	"runtime"
	"strconv"
	"strings"
	"sync/atomic"
	"time"

	_ "github.com/pingcap/tidb/pkgdep"

	"github.com/opentracing/opentracing-go"
	"github.com/pingcap/errors"
        .....
```

Comment (tiancaiamao) [reply perfschema/init.go]: 1. Not in this PR
2. I'm not sure whether it works
@lysu

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -868,63 +870,64 @@ func (d *ddl) CreateTableWithLike(ctx sessionctx.Context, ident, referIdent ast.
...
-func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err error) {
-	ident := ast.Ident{Schema: s.Table.Schema, Name: s.Table.Name}
-	if s.ReferTable != nil {
-		referIdent := ast.Ident{Schema: s.ReferTable.Schema, Name: s.ReferTable.Name}
-		return d.CreateTableWithLike(ctx, ident, referIdent, s.IfNotExists)
+// BuildTableInfoFromSQL builds model.TableInfo from a SQL statement.
+// The SQL string should be a create table statement.
+// Don't use this function to build a partitioned table.
+func BuildTableInfoFromSQL(sql string) (*model.TableInfo, error) {
+	p := parser.New()
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply ddl/ddl_api.go]: Change the argument to an AST ?

[the remark this answers is not in the document]
Comment (tiancaiamao) [reply ddl/ddl_api.go]: Done

**Comment by tiancaiamao** [comment]
PTAL @lysu

**Comment by tiancaiamao** [comment]
PTAL @jackysp @winkyao

**Comment by tiancaiamao** [comment]
PTAL @winkyao

**Comment by tiancaiamao** [comment]
PTAL @winkyao
